man/ndbxfrm.1
'\" t
.\"     Title: ndbxfrm
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 06/25/2026
.\"    Manual: MySQL Database System
.\"    Source: MySQL 9.7
.\"  Language: English
.\"
.TH "NDBXFRM" "1" "06/25/2026" "MySQL 9\&.7" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ndbxfrm \- compress, decompress, encrypt, and decrypt files created by NDB Cluster
.SH "SYNOPSIS"
.HP \w'\fBndbxfrm\ \fR\fB\fIoptions\fR\fR\fB\ \fR\fB\fIfilenames\fR\fR\ 'u
\fBndbxfrm \fR\fB\fIoptions\fR\fR\fB \fR\fB\fIfilenames\fR\fR
.SH "DESCRIPTION"
.PP
The
\fBndbxfrm\fR
utility can be used to decompress, decrypt, and output information about files created by NDB Cluster that are compressed, encrypted, or both\&. It can also be used to compress or encrypt files\&.
.PP
Usage
.sp
.if n \{\
.RS 4
.\}
.nf
ndbxfrm \-\-info \fIfile\fR[ \fIfile\fR \&.\&.\&.]
ndbxfrm \-\-compress \fIinput_file\fR \fIoutput_file\fR
ndbxfrm \-\-decrypt\-password=\fIpassword\fR \fIinput_file\fR \fIoutput_file\fR
ndbxfrm [\-\-encrypt\-ldf\-iter\-count=#] \-\-encrypt\-password=\fIpassword\fR \fIinput_file\fR \fIoutput_file\fR
.fi
.if n \{\
.RE
.\}
.PP
\fIinput_file\fR
and
\fIoutput_file\fR
cannot be the same file\&.
Options
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-compress\fR,
\fB\-c\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--compress
T}
.TE
.sp 1
Compresses the input file, using the same compression method as is used for compressing NDB Cluster backups, and writes the output to an output file\&. To decompress a compressed
NDB
backup file that is not encrypted, it is necessary only to invoke
\fBndbxfrm\fR
using the names of the compressed file and an output file (with no options required)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-decrypt\-key=\fR\fB\fIkey\fR\fR,
\fB\-K\fR
\fIkey\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--decrypt-key=key
T}
.TE
.sp 1
Decrypts a file encrypted by
NDB
using the supplied key\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
This option cannot be used together with
\fB\-\-decrypt\-password\fR\&.
.sp .5v
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-decrypt\-key\-from\-stdin\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--decrypt-key-from-stdin
T}
.TE
.sp 1
Decrypts a file encrypted by
NDB
using the key supplied from
stdin\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-decrypt\-password=\fR\fB\fIpassword\fR\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--decrypt-password=password
T}
T{
Type
T}:T{
String
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Decrypts a file encrypted by
NDB
using the password supplied\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
This option cannot be used together with
\fB\-\-decrypt\-key\fR\&.
.sp .5v
.RE
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-decrypt\-password\-from\-stdin[=TRUE|FALSE]\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--decrypt-password-from-stdin
T}
.TE
.sp 1
Decrypts a file encrypted by
NDB, using a password supplied from standard input\&. This is similar to entering a password after invoking
\fBmysql\fR
\fB\-\-password\fR
with no password following the option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-extra\-file\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--defaults-extra-file=path
T}
T{
Type
T}:T{
String
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Read given file after global files are read\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-file\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--defaults-file=path
T}
T{
Type
T}:T{
String
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Read default options from given file only\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-group\-suffix\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--defaults-group-suffix=string
T}
T{
Type
T}:T{
String
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Also read groups with
CONCAT(\fIgroup\fR, \fIsuffix\fR)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-detailed\-info\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--encrypt-block-size=#
T}
T{
Type
T}:T{
Boolean
T}
T{
Default Value
T}:T{
FALSE
T}
.TE
.sp 1
Print out file information like
\fB\-\-info\fR, but include the file\*(Aqs header and trailer\&.
.sp
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBndbxfrm \-\-detailed\-info S0\&.sysfile\fR
File=/var/lib/cluster\-data/ndb_7_fs/D1/NDBCNTR/S0\&.sysfile, compression=no, encryption=yes
header: {
  fixed_header: {
    magic: {
      magic: { 78, 68, 66, 88, 70, 82, 77, 49 },
      endian: 18364758544493064720,
      header_size: 32768,
      fixed_header_size: 160,
      zeros: { 0, 0 }
    },
    flags: 73728,
    flag_extended: 0,
    flag_zeros: 0,
    flag_file_checksum: 0,
    flag_data_checksum: 0,
    flag_compress: 0,
    flag_compress_method: 0,
    flag_compress_padding: 0,
    flag_encrypt: 18,
    flag_encrypt_cipher: 2,
    flag_encrypt_krm: 1,
    flag_encrypt_padding: 0,
    flag_encrypt_key_selection_mode: 0,
    dbg_writer_ndb_version: 524320,
    octets_size: 32,
    file_block_size: 32768,
    trailer_max_size: 80,
    file_checksum: { 0, 0, 0, 0 },
    data_checksum: { 0, 0, 0, 0 },
    zeros01: { 0 },
    compress_dbg_writer_header_version: { \&.\&.\&. },
    compress_dbg_writer_library_version: { \&.\&.\&. },
    encrypt_dbg_writer_header_version: { \&.\&.\&. },
    encrypt_dbg_writer_library_version: { \&.\&.\&. },
    encrypt_key_definition_iterator_count: 100000,
    encrypt_krm_keying_material_size: 32,
    encrypt_krm_keying_material_count: 1,
    encrypt_key_data_unit_size: 32768,
    encrypt_krm_keying_material_position_in_octets: 0,
  },
  octets: {
     102, 68, 56, 125, 78, 217, 110, 94, 145, 121, 203, 234, 26, 164, 137, 180,
     100, 224, 7, 88, 173, 123, 209, 110, 185, 227, 85, 174, 109, 123, 96, 156,
  }
}
trailer: {
  fixed_trailer: {
    flags: 48,
    flag_extended: 0,
    flag_zeros: 0,
    flag_file_checksum: 0,
    flag_data_checksum: 3,
    data_size: 512,
    file_checksum: { 0, 0, 0, 0 },
    data_checksum: { 226, 223, 102, 207 },
    magic: {
      zeros: { 0, 0 }
      fixed_trailer_size: 56,
      trailer_size: 32256,
      endian: 18364758544493064720,
      magic: { 78, 68, 66, 88, 70, 82, 77, 49 },
    },
  }
}
.fi
.if n \{\
.RE
.\}
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-encrypt\-block\-size=\fR\fB\fI#\fR\fR
.TS
allbox tab(:);
lB l
lB l
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--encrypt-block-size=#
T}
T{
Type
T}:T{
Integer
T}
T{
Default Value
T}:T{
0
T}
T{
Minimum Value
T}:T{
0
T}
T{
Maximum Value
T}:T{
2147483647
T}
.TE
.sp 1
Size of input data chunks that are encrypted as a unit\&. Used with XTS; set to
0
(the default) for CBC mode\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-encrypt\-cipher=\fR\fB\fI#\fR\fR
.TS
allbox tab(:);
lB l
lB l
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--encrypt-cipher=#
T}
T{
Type
T}:T{
Integer
T}
T{
Default Value
T}:T{
1
T}
T{
Minimum Value
T}:T{
0
T}
T{
Maximum Value
T}:T{
2147483647
T}
.TE
.sp 1
Cipher used for encryption\&. Set to
1
for CBC mode (the default), or
2
for XTS\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-encrypt\-kdf\-iter\-count=\fR\fB\fI#\fR\fR,
\fB\-k \fR\fB\fI#\fR\fR
.TS
allbox tab(:);
lB l
lB l
lB l
lB l
lB l.
T{

ChangeLog
2010-03-18	The InnoDB Team

	* CMakeLists.txt:
	Fix Bug#52102 InnoDB Plugin shows performance drop compared to
	InnoDB (Windows)

2010-03-18	The InnoDB Team

	* buf0buf.ic:
	When comparing the time of the first access to a block against
	innodb_old_blocks_time, use 32-bit arithmetics. The comparison was
	incorrect on 64-bit systems.

2010-03-11	The InnoDB Team

	* buf0buf.h, buf0buf.ic:
	Fix and clarify the latching of some buf_block_t members.
	Note that check_index_page_at_flush is not protected by any mutex.
	Note and assert that lock_hash_val is protected by the rw-latch.

2010-03-10	The InnoDB Team

	* trx/trx0sys.c:
	Fix Bug#51653 outdated reference to set-variable

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb_bug21704.result,
	mysql-test/innodb_bug47621.result, mysql-test/innodb_bug47621.test:
	Fix Bug#47621 MySQL and InnoDB data dictionaries will become out of
	sync when renaming columns

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc:
	Fix Bug#51356 Many Valgrind errors in error messages
	with concurrent DDL

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, handler/handler0alter.cc,
	mysql-test/innodb_bug51378.result, mysql-test/innodb_bug51378.test:
	Fix Bug#51378 Init 'ref_length' to correct value, in case an out
	of bound MySQL primary_key

2010-03-10	The InnoDB Team

	* log/log0recv.c:
	Remove a bogus assertion about page numbers exceeding 0x90000000
	in the redo log. Abort when encountering a corrupted redo log
	record, unless innodb_force_recovery is set.

2010-03-09	The InnoDB Team

	* handler/ha_innodb.cc:
	Make SHOW ENGINE INNODB MUTEX STATUS display SUM(os_waits)
	for the buffer pool block mutexes and locks.

2010-03-08	The InnoDB Team

	* fil/fil0fil.c:
	Fix ALTER TABLE ... IMPORT TABLESPACE of compressed tables.

2010-03-03	The InnoDB Team

	* handler/handler0alter.cc, innodb-index.result, innodb-index.test,
	innodb.result, innodb.test:
	Disallow a duplicate index name when creating an index.

2010-02-11	The InnoDB Team

	* include/mem0mem.h, include/mem0mem.ic, mem/mem0mem.c:
	Fix Bug#49535 Available memory check slows down crash
	recovery tens of times

2010-02-09	The InnoDB Team

	* buf/buf0buf.c:
	Fix Bug#38901 InnoDB logs error repeatedly when trying to load
	page into buffer pool

2010-02-09	The InnoDB Team

	* srv/srv0srv.c:
	Let the master thread sleep if the amount of work to be done is
	calibrated as taking less than a second.

2010-02-04	The InnoDB Team

	* btr/btr0btr.c, btr/btr0cur.c, btr/btr0pcur.c, buf/buf0buf.c,
	include/btr0btr.h, include/btr0cur.h, include/btr0pcur.h,
	include/btr0pcur.ic, include/buf0buf.h, row/row0ins.c, row/row0sel.c:
	Pass the file name and line number of the caller of the
	b-tree cursor functions to the buffer pool requests, in order
	to make the latch diagnostics more accurate.

2010-02-03	The InnoDB Team

	* lock/lock0lock.c:
	Fix Bug#49001 SHOW INNODB STATUS deadlock info incorrect
	when deadlock detection aborts

2010-02-03	The InnoDB Team

	* buf/buf0lru.c:
	Fix Bug#35077 Very slow DROP TABLE (ALTER TABLE, OPTIMIZE TABLE)
	on compressed tables

2010-02-03	The InnoDB Team

	* handler/ha_innodb.cc, include/row0mysql.h, row/row0mysql.c:
	Clean up CHECK TABLE error handling.

2010-02-01	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb-autoinc.test,
	mysql-test/innodb-autoinc.result,
	mysql-test/innodb-autoinc-44030.test,
	mysql-test/innodb-autoinc-44030.result:
	Fix Bug#49497 Error 1467 (ER_AUTOINC_READ_FAILED) on inserting
	a negative value

2010-01-27	The InnoDB Team

	* include/row0mysql.h, log/log0recv.c, row/row0mysql.c:
	Drop temporary tables at startup.
	This addresses the third aspect of
	Bug#41609 Crash recovery does not work for InnoDB temporary tables.

2010-01-21	The InnoDB Team

	* buf/buf0buf.c:
	Do not merge buffered inserts to compressed pages before
	the redo log has been applied in crash recovery.

2010-01-13	The InnoDB Team

	* row/row0sel.c:
	On the READ UNCOMMITTED isolation level, do not attempt to access
	a clustered index record that has been marked for deletion. The
	built-in InnoDB in MySQL 5.1 and earlier would attempt to retrieve
	a previous version of the record in this case.

2010-01-13	The InnoDB Team

	* buf/buf0buf.c:
	When disabling the adaptive hash index, check the block state
	before checking block->is_hashed, because the latter may be
	uninitialized right after server startup.

2010-01-12	The InnoDB Team


storage/ndb/tools/ndbxfrm.cpp
/*
   Copyright (c) 2020, 2026, Oracle and/or its affiliates.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is designed to work with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have either included with
   the program or referenced in the documentation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

#include <cstdio>
#include <cstring>
#include "util/require.h"

#include "kernel/signaldata/FsOpenReq.hpp"
#include "my_getopt.h"
#include "portlib/ndb_file.h"
#include "util/ndb_ndbxfrm1.h"
#include "util/ndb_openssl_evp.h"
#include "util/ndb_opts.h"
#include "util/ndbxfrm_buffer.h"
#include "util/ndbxfrm_file.h"
#include "util/ndbxfrm_iterator.h"

using byte = unsigned char;

static int g_compress = 0;

static ndb_password_state opt_decrypt_password_state("decrypt", nullptr);
static ndb_password_option opt_decrypt_password(opt_decrypt_password_state);
static ndb_password_from_stdin_option opt_decrypt_password_from_stdin(
    opt_decrypt_password_state);

static ndb_password_state opt_encrypt_password_state("encrypt", nullptr);
static ndb_password_option opt_encrypt_password(opt_encrypt_password_state);
static ndb_password_from_stdin_option opt_encrypt_password_from_stdin(
    opt_encrypt_password_state);

static ndb_key_state opt_decrypt_key_state("decrypt", nullptr);
static ndb_key_option opt_decrypt_key(opt_decrypt_key_state);
static ndb_key_from_stdin_option opt_decrypt_key_from_stdin(
    opt_decrypt_key_state);

static ndb_key_state opt_encrypt_key_state("encrypt", nullptr);
static ndb_key_option opt_encrypt_key(opt_encrypt_key_state);
static ndb_key_from_stdin_option opt_encrypt_key_from_stdin(
    opt_encrypt_key_state);

static int g_info = 0;
static int g_detailed_info = 0;
static int g_encrypt_block_size = 0;
static int g_encrypt_cipher = ndb_ndbxfrm1::cipher_cbc;
static int g_encrypt_kdf_iter_count =
    -1;  // ndb_openssl_evp::DEFAULT_KDF_ITER_COUNT;
static int g_file_block_size = 512;
#if defined(TODO_READ_REVERSE)
static int g_read_reverse = 0;
#endif

// clang-format off
static struct my_option my_long_options[] =
{
  NdbStdOpt::usage,
  NdbStdOpt::help,
  NdbStdOpt::version,

  // Specific options
  { "compress", 'c', "Compress file",
    &g_compress, nullptr, nullptr, GET_BOOL, NO_ARG,
    0, 0, 0, nullptr, 0, nullptr },
  { "decrypt-key", NDB_OPT_NOSHORT, "Decryption key",
    nullptr, nullptr, nullptr, GET_PASSWORD, OPT_ARG,
    0, 0, 0, nullptr, 0, &opt_decrypt_key},
  { "decrypt-key-from-stdin", NDB_OPT_NOSHORT, "Decryption key",
    &opt_decrypt_key_from_stdin.opt_value, nullptr, nullptr, GET_BOOL, NO_ARG,
    0, 0, 0, nullptr, 0, &opt_decrypt_key_from_stdin},
  { "decrypt-password", NDB_OPT_NOSHORT, "Decryption password",
    nullptr, nullptr, nullptr, GET_PASSWORD, OPT_ARG,
    0, 0, 0, nullptr, 0, &opt_decrypt_password},
  { "decrypt-password-from-stdin", NDB_OPT_NOSHORT, "Decryption password",
    &opt_decrypt_password_from_stdin.opt_value, nullptr, nullptr, GET_BOOL, NO_ARG,
    0, 0, 0, nullptr, 0, &opt_decrypt_password_from_stdin},
  { "encrypt-block-size", NO_ARG,
    "Size of input data chunks that are encrypted as an unit. Used with XTS, "
    "zero for CBC mode.",
    &g_encrypt_block_size, nullptr, nullptr, GET_INT, REQUIRED_ARG,
    0, 0, INT_MAX, nullptr, 0, nullptr },
  { "encrypt-cipher", NO_ARG, "Encrypt cipher: CBC(1), XTS(2).",
    &g_encrypt_cipher, nullptr, nullptr, GET_INT, REQUIRED_ARG,
    ndb_ndbxfrm1::cipher_cbc, 0, INT_MAX, nullptr, 0, nullptr },
  { "encrypt-kdf-iter-count", 'k', "Iteration count to used in key definition",
    &g_encrypt_kdf_iter_count, nullptr, nullptr, GET_INT, REQUIRED_ARG,
    ndb_openssl_evp::DEFAULT_KDF_ITER_COUNT, 0, INT_MAX, nullptr, 0, nullptr },
  { "encrypt-key", NDB_OPT_NOSHORT, "Encryption key",
    nullptr, nullptr, nullptr, GET_PASSWORD, OPT_ARG,
    0, 0, 0, nullptr, 0, &opt_encrypt_key},
  { "encrypt-key-from-stdin", NDB_OPT_NOSHORT, "Encryption key",
    &opt_encrypt_key_from_stdin.opt_value, nullptr, nullptr,
    GET_BOOL, NO_ARG, 0, 0, 0, nullptr, 0, &opt_encrypt_key_from_stdin},
  { "encrypt-password", NDB_OPT_NOSHORT, "Encryption password",
    nullptr, nullptr, nullptr, GET_PASSWORD, OPT_ARG,
    0, 0, 0, nullptr, 0, &opt_encrypt_password},
  { "encrypt-password-from-stdin", NDB_OPT_NOSHORT, "Encryption password",
    &opt_encrypt_password_from_stdin.opt_value, nullptr, nullptr,
    GET_BOOL, NO_ARG, 0, 0, 0, nullptr, 0, &opt_encrypt_password_from_stdin},
  { "file-block-size", NO_ARG, "File block size.",
    &g_file_block_size, nullptr, nullptr, GET_INT, REQUIRED_ARG,
    512, 0, INT_MAX, nullptr, 0, nullptr },
  { "info", 'i', "Print info about file",
    &g_info, nullptr, nullptr, GET_BOOL, NO_ARG,
    0, 0, 0, nullptr, 0, nullptr },
    { "detailed-info", NDB_OPT_NOSHORT, "Print info about file including file header and trailer",
    &g_detailed_info, nullptr, nullptr, GET_BOOL, NO_ARG,
    0, 0, 0, nullptr, 0, nullptr },
#if defined(TODO_READ_REVERSE)
  { "read-reverse", 'R', "Read file in reverse",
    &g_read_reverse, nullptr, nullptr, GET_BOOL, NO_ARG,
    0, 0, 0, nullptr, 0, nullptr },
#endif
  NdbStdOpt::end_of_options
};
// clang-format on

static const char *load_defaults_groups[] = {"ndbxfrm", nullptr};

static int dump_info(const char name[], bool print_header_and_trailer);
static int copy_file(const char src[], const char dst[]);

int main(int argc, char *argv[]) {
  NDB_INIT(argv[0]);
  Ndb_opts opts(argc, argv, my_long_options, load_defaults_groups);
  if (opts.handle_options()) return 2;

#if defined(TODO_READ_REVERSE)
  bool do_write = (g_compress || (opt_encrypt_password != nullptr));

  if (g_read_reverse && do_write) {
    fprintf(stderr,
            "Error: Writing with encryption (--encrypt-password) or "
            "compression (--compress) not allowed when reading in reverse "
            "(--read-reverse).\n");
    return 1;
  }

  if (g_info || g_detailed_info) && (do_write || g_read_reverse))
  {
      fprintf(stderr,
              "Error: Writing (--encrypt-password, --compress) or reading "
              "(--read-reverse) file not allowed in connection with --info.\n");
      return 1;
    }
#endif

  if (ndb_option::post_process_options()) {
    BaseString err_msg = opt_decrypt_key_state.get_error_message();
    if (!err_msg.empty()) {
      fprintf(stderr, "Error: %s\n", err_msg.c_str());
    }
    err_msg = opt_decrypt_password_state.get_error_message();
    if (!err_msg.empty()) {
      fprintf(stderr, "Error: %s\n", err_msg.c_str());
    }
    err_msg = opt_encrypt_key_state.get_error_message();
    if (!err_msg.empty()) {
      fprintf(stderr, "Error: %s\n", err_msg.c_str());
    }
    err_msg = opt_encrypt_password_state.get_error_message();
    if (!err_msg.empty()) {
      fprintf(stderr, "Error: %s\n", err_msg.c_str());
    }
    return 2;
  }
  if (opt_decrypt_key_state.get_key() != nullptr &&
      opt_decrypt_password_state.get_password() != nullptr) {
    fprintf(stderr, "Error: Both decrypt key and decrypt password is set.\n");
    return 2;
  }
  if (opt_encrypt_key_state.get_key() != nullptr &&
      opt_encrypt_password_state.get_password() != nullptr) {
    fprintf(stderr, "Error: Both encrypt key and encrypt password is set.\n");
    return 2;
  }
  if ((opt_decrypt_key_state.get_key() != nullptr ||
       opt_encrypt_key_state.get_key() != nullptr) &&
      !ndb_openssl_evp::is_aeskw256_supported()) {
    fprintf(stderr,
            "Error: decrypt and encrypt key options requires OpenSSL 1.0.2 "
            "or newer.\n");
    return 2;
  }

  if (g_detailed_info) {
    for (int argi = 0; argi < argc; argi++) {
      dump_info(argv[argi], true);
    }
    return 0;
  }

  if (g_info) {
    for (int argi = 0; argi < argc; argi++) {
      dump_info(argv[argi], false);
    }
    return 0;
  }

  if (g_file_block_size < 0) {
    fprintf(stderr, "Error: file_block_size %d can not be negative.\n",
            g_file_block_size);
    return 1;
  }

  if (argc != 2) {
    fprintf(stderr, "Error: Need one source file and one destination file.\n");
    return 1;
  }

  ndb_openssl_evp::library_init();

  int rc = copy_file(argv[0], argv[1]);

  ndb_openssl_evp::library_end();

  return rc;
}

int dump_info(const char name[], bool print_header_and_trailer) {
  ndb_file file;
  ndbxfrm_file xfrm;
  int r;

  r = file.open(name, FsOpenReq::OM_READONLY);
  if (r == -1) {
    fprintf(stderr, "Error: Could not open file '%s' for read.\n", name);
    return 1;
  }
  ndb_ndbxfrm1::header header;
  ndb_ndbxfrm1::trailer trailer;
  r = xfrm.read_header_and_trailer(file, header, trailer);
  if (r == -1) {
    fprintf(stderr, "Error: Could not read file '%s'.\n", name);
    return 1;
  }
  require(r == 0);

  Uint32 cipher = 0;
  header.get_encryption_cipher(&cipher);
  bool is_compressed = (header.get_compression_method() != 0);
  bool is_encrypted = (cipher != 0);
  printf("File=%s, compression=%s, encryption=%s\n", name,
         is_compressed ? "yes" : "no", is_encrypted ? "yes" : "no");

  if (print_header_and_trailer) {
    header.printf(stdout);
    trailer.printf(stdout);
  }

  file.close();
  return 0;
}

int copy_file(const char src[], const char dst[]) {
  ndb_file src_file;
  ndb_file dst_file;

  int r;

  if (dst_file.create(dst) != 0) {
    fprintf(stderr, "Error: Could not create file '%s'.\n", dst);
    perror(dst);
    return 1;  // File exists?
  }

  r = src_file.open(src, FsOpenReq::OM_READONLY);
  if (r == -1) {
    fprintf(stderr, "Error: Could not open file '%s' for read.\n", src);
    perror(src);
    dst_file.remove(dst);
    return 1;
  }

  r = dst_file.open(dst, FsOpenReq::OM_WRITEONLY);
  if (r == -1) {
    fprintf(stderr, "Error: Could not open file '%s' for write.\n", dst);
    perror(dst);
    src_file.close();
    dst_file.remove(dst);
    return 1;
  }

  ndbxfrm_file src_xfrm;
  ndbxfrm_file dst_xfrm;

  const byte *src_pwd_key =
      opt_decrypt_key_state.get_key() != nullptr
          ? opt_decrypt_key_state.get_key()
          : reinterpret_cast<const byte *>(
                opt_decrypt_password_state.get_password());
  const size_t src_pwd_key_len =
      opt_decrypt_key_state.get_key() != nullptr
          ? opt_decrypt_key_state.get_key_length()
          : opt_decrypt_password_state.get_password_length();
  r = src_xfrm.open(src_file, src_pwd_key, src_pwd_key_len);
  if (r == -1) {
    fprintf(stderr, "Error: Can not read file %s, bad password or key?\n", src);
    src_file.close();
    dst_file.close();
    dst_file.remove(dst);
    return 1;
  }

  require(g_file_block_size >= 0);
  size_t file_block_size = g_file_block_size;

  const byte *dst_pwd_key =
      opt_encrypt_key_state.get_key() != nullptr
          ? opt_encrypt_key_state.get_key()
          : reinterpret_cast<const byte *>(
                opt_encrypt_password_state.get_password());
  const size_t dst_pwd_key_len =
      opt_encrypt_key_state.get_key() != nullptr
          ? opt_encrypt_key_state.get_key_length()
          : opt_encrypt_password_state.get_password_length();
  const Uint64 file_size = src_file.get_size();
  r = dst_xfrm.create(dst_file, g_compress, dst_pwd_key, dst_pwd_key_len,
                      g_encrypt_kdf_iter_count, g_encrypt_cipher,
                      -1 /* key count */, g_encrypt_block_size, file_block_size,
                      file_size, true, true);
  if (r != 0) {
    fprintf(stderr, "Error: Can not initialize file %s.\n", dst);
    src_xfrm.close(true);
    src_file.close();
    dst_file.close();
    dst_file.remove(dst);
    return 1;
  }

  // Copy data

  ndbxfrm_buffer buffer;
  buffer.init();
  for (;;) {
    ndbxfrm_input_iterator wr_it = buffer.get_input_iterator();
    if (dst_xfrm.write_forward(&wr_it) == -1) {
      fprintf(stderr, "Error: Can not write file %s.\n", dst);
      r = 2;  // write failure
      break;
    }
    buffer.update_read(wr_it);
    buffer.rebase(0);

    if (buffer.last() && buffer.read_size() == 0) {
      break;  // All read and written
    }

    ndbxfrm_output_iterator rd_it = buffer.get_output_iterator();
    if (src_xfrm.read_forward(&rd_it) == -1) {
      if (src_xfrm.is_encrypted()) {
        fprintf(stderr, "Error: Can not read file %s, bad password or key?\n",
                src);
      } else {
        fprintf(stderr, "Error: Can not read file %s.\n", src);
      }
      r = 2;  // read failure
      break;
    }
    buffer.update_write(rd_it);
  }

  if (r != 0) {
    src_xfrm.close(true);
    src_file.close();
  } else {
    r = src_xfrm.close(false);
    if (r != 0) {
      if (src_xfrm.is_encrypted()) {
        fprintf(stderr, "Error: Can not read file %s, bad password or key?\n",
                src);
      } else {
        fprintf(stderr, "Error: Can not read file %s.\n", src);
      }
      r = 2;  // read failure
    }
    if (src_file.close() != 0 && r == 0) {
      fprintf(stderr, "Error: Can not read file %s.\n", src);
      r = 2;  // read failure
    }
  }
  dst_xfrm.close((r != 0));

  dst_file.sync();
  dst_file.close();

  if (r != 0) {
    dst_file.remove(dst);
  }

  return r;
}
